File: cleanclaw/projectmap/store.py
import json
import os
from pathlib import Path

import faiss
import numpy as np


def _index_path(store_dir: str, layer: str) -> str:
    return str(Path(store_dir) / f"{layer}.index")


def _meta_path(store_dir: str, layer: str) -> str:
    return str(Path(store_dir) / f"{layer}.json")


def load_table(store_dir: str, layer: str) -> tuple[faiss.Index | None, list[dict]]:
    """Load FAISS index and metadata sidecar for a layer. Returns (index, rows)."""
    ip = _index_path(store_dir, layer)
    mp = _meta_path(store_dir, layer)
    if not os.path.exists(ip) or not os.path.exists(mp):
        return None, []
    index = faiss.read_index(ip)
    with open(mp, encoding="utf-8") as f:
        rows = json.load(f)
    return index, rows


def save_table(store_dir: str, layer: str, rows: list[dict], vectors: list[list[float]]) -> None:
    """Build a flat FAISS index from vectors and save with metadata sidecar."""
    os.makedirs(store_dir, exist_ok=True)
    if not vectors:
        return
    arr = np.array(vectors, dtype="float32")
    faiss.normalize_L2(arr)
    index = faiss.IndexFlatIP(arr.shape[1])
    index.add(arr)
    faiss.write_index(index, _index_path(store_dir, layer))
    with open(_meta_path(store_dir, layer), "w", encoding="utf-8") as f:
        json.dump(rows, f, indent=2)


def remove_file_rows(store_dir: str, layer: str, full_path: str) -> tuple[list[dict], list[int]]:
    """Return rows and original positions with full_path removed."""
    _, all_rows = load_table(store_dir, layer)
    kept_rows = [r for r in all_rows if r.get("full_path") != full_path]
    kept_positions = [i for i, r in enumerate(all_rows) if r.get("full_path") != full_path]
    return kept_rows, kept_positions


def query_table(store_dir: str, layer: str, query_vector: list[float], top_k: int = 10) -> list[dict]:
    """Return top-k metadata rows most similar to query_vector."""
    index, rows = load_table(store_dir, layer)
    if index is None or not rows:
        return []
    arr = np.array([query_vector], dtype="float32")
    faiss.normalize_L2(arr)
    distances, indices = index.search(arr, min(top_k, len(rows)))
    return [rows[i] for i in indices[0] if i < len(rows)]

File: cleanclaw/projectmap/store.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";

export interface MethodRow {
  method_name: string;
  signature: string;
  output: string;
  filename: string;
  full_path: string;
  metadata: string;
  algorithm: string;
}

export interface MiscRow {
  filename: string;
  full_path: string;
  purpose: string;
  related_layer: string;
}

export type StoreRow = MethodRow | MiscRow;

function indexPath(storeDir: string, layer: string): string {
  return join(storeDir, `${layer}.vectors.json`);
}

function metaPath(storeDir: string, layer: string): string {
  return join(storeDir, `${layer}.json`);
}

export interface LoadResult {
  vectors: number[][];
  rows: StoreRow[];
}

export function loadTable(storeDir: string, layer: string): LoadResult {
  const ip = indexPath(storeDir, layer);
  const mp = metaPath(storeDir, layer);
  if (!existsSync(ip) || !existsSync(mp)) {
    return { vectors: [], rows: [] };
  }
  const vectors = JSON.parse(readFileSync(ip, "utf-8")) as number[][];
  const rows = JSON.parse(readFileSync(mp, "utf-8")) as StoreRow[];
  return { vectors, rows };
}

export function saveTable(
  storeDir: string,
  layer: string,
  rows: StoreRow[],
  vectors: number[][]
): void {
  if (vectors.length === 0) return;
  mkdirSync(storeDir, { recursive: true });
  writeFileSync(indexPath(storeDir, layer), JSON.stringify(vectors, null, 2), "utf-8");
  writeFileSync(metaPath(storeDir, layer), JSON.stringify(rows, null, 2), "utf-8");
}

export function removeFileRows(
  storeDir: string,
  layer: string,
  fullPath: string
): { rows: StoreRow[]; positions: number[] } {
  const { rows: allRows } = loadTable(storeDir, layer);
  const rows: StoreRow[] = [];
  const positions: number[] = [];
  for (let i = 0; i < allRows.length; i++) {
    if ((allRows[i] as { full_path?: string }).full_path !== fullPath) {
      rows.push(allRows[i]);
      positions.push(i);
    }
  }
  return { rows, positions };
}

function dotProduct(a: number[], b: number[]): number {
  let sum = 0;
  for (let i = 0; i < a.length; i++) {
    sum += a[i] * b[i];
  }
  return sum;
}

function l2Norm(v: number[]): number {
  return Math.sqrt(v.reduce((s, x) => s + x * x, 0));
}

function cosineSimilarity(a: number[], b: number[]): number {
  const na = l2Norm(a);
  const nb = l2Norm(b);
  if (na === 0 || nb === 0) return 0;
  return dotProduct(a, b) / (na * nb);
}

export function queryTable(
  storeDir: string,
  layer: string,
  queryVector: number[],
  topK = 10
): StoreRow[] {
  const { vectors, rows } = loadTable(storeDir, layer);
  if (vectors.length === 0) return [];

  const scored = vectors.map((v, i) => ({ score: cosineSimilarity(queryVector, v), i }));
  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, topK).map((s) => rows[s.i]);
}
